### PR TITLE
i18n: Fix source root in Gettext targets for subprojects

### PR DESCRIPTION
**Pull request by dpward**
Gettext should search for input files relative to the (sub)project source root, not the global source root.

### CODE REVIEW

File: mesonbuild/modules/i18n.py [hunk 1]
```diff
@@ -256,10 +256,18 @@ def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -
 
         extra_arg = '--extra-args=' + '@@'.join(extra_args) if extra_args else None
 
+        if state.subproject:
+            source_root = path.join(state.environment.source_dir, self.interpreter.subproject_dir, state.subproject)
```
Comment (eli-schwartz) [inline mesonbuild/modules/i18n.py:260]: The name of the subproject is not necessarily the name of its directory.

IIRC you can use interpreter.root_subdir for this.

**Review by eli-schwartz — COMMENTED** [review]
Thanks for looking into this.

File: mesonbuild/modules/i18n.py (hunk 1 above)
Comment (eli-schwartz) [reply mesonbuild/modules/i18n.py]: That's... entirely possible lol.

If it's wrong for one location it should be wrong for both of them. But documentation builds aren't all that more common than maintainer builds when dealing with subprojects.

Comment (eli-schwartz) [reply mesonbuild/modules/i18n.py]: Yep, you can confirm this if you try to build a hotdoc subproject and look at `subprojects/foobar-1.0/*-doc.json` and observe the `"gi_c_source_roots"` key.

It does indeed need fixing, want to fix it as a followup commit in this PR?

Comment (eli-schwartz) [reply mesonbuild/modules/i18n.py]: Wait, lol. Even worse. If it is NOT a subproject, this caused `"gi_c_source_roots"` to be e.g.

```
[
    "path/to/meson/docs/subprojects/",
    "path/to/meson/docs/builddir/subprojects/"
]
```

Which seems excessively broken...

Comment (eli-schwartz) [reply mesonbuild/modules/i18n.py]: @thiblahute This seems probably broken from commit c6012e3cc728982797e794fd28b89f1ff3bb25cc and b4347ca4a98b148ba7f7908c3d96f393294df122?

**Comment by xclaesse** [comment]
Seems worth adding `state.root_subdir`

**Comment by eli-schwartz** [comment]
Sure, incremental improvements and all that. :)

More importantly, this reduces overly long lines even more than the PR already did!

**Comment by eli-schwartz** [comment]
I never have the faintest clue what codecov is doing. :shrug: 

Anyway, seems like the original author of this code isn't available ATM to weigh in on the original intention. But the fix sure appears correct. :)

**Pull request merged**